src/seismometer/data/pandas_helpers.py

```diff
@@ -153,11 +175,132 @@ def infer_label(dataframe: pd.DataFrame, label_col: str, time_col: str) -> pd.Da
+    count_cols = list(event_name_map.values())
+    left[count_cols] = left[count_cols].fillna(0) #Fill any missing counts for rows that didn't have any events
+    
+    return left
+
+
+def _merge_with_strategy(
+    predictions: pd.DataFrame,
+    one_event: pd.DataFrame,
+    pks: list[str],
+    *,
+    pred_ref: str = "Time",
+    event_ref: str = "Time",
+    event_display: str = "an event",
+    merge_strategy: str = "forward",
+) -> pd.DataFrame:
+    """
+    Merges the right frame into the left based on a set of exact match primary keys and merge strategy. 
+
+    Parameters
+    ----------
+    predictions : pd.DataFrame
+        The left frame, usually of predictions. Assumed to be sorted by time.
+    one_event : pd.DataFrame
+        The right frame to merge, assumed to be of events. Assumed to be sorted by time if applicable.
+    pks : list[str]
+        The list of columns to require exact matches during the merge.
+    pred_ref : str, optional
+        The column in the left (prediction) frame to use as a reference point in the distance match, by default 'Time'.
+    event_ref : str, optional
+        The column in the right (event) frame to use reference in the distance match, by default 'Time'.
+    event_display : str, optional
+        The name of the event to display in warning messages, by default "an event"
+    merge_strategy : str
+        The method to use when merging the event data, by default 'forward'.
+
+    Returns
+    -------
+    pd.DataFrame
+        The merged dataframe.
+    """
+    try:
+        ct_times = one_event[event_ref].notna().sum()
+
+        #If there are no times in the event frame, merge the first row for each group
+        if ct_times == 0:
+                #Set the filtered frame to the first row for each group and throw a value error which is passed before merging.
+                one_event_filtered = one_event.groupby(pks).first()
+                raise ValueError(f"No times found for {event_display}, merging first row for each group.")
+
+        if ct_times != len(one_event.index):
+            logger.warning(f"Inconsistent event times for {event_display}, only considering events with times.")
+            one_event = one_event.dropna(subset=[event_ref])
+
+        if merge_strategy=="forward" or merge_strategy=="nearest":
+            return pd.merge_asof(
+                predictions,
+                one_event.dropna(subset=[event_ref]),
+                left_on=pred_ref,
+                right_on=event_ref,
+                by=pks,
+                direction=merge_strategy,
+            )
+
+        #If there's multiple events with matching event_ref vals, idxmax and idxmin will return the first row.
+        #So we sort the index before grabbing the value to default to the first and last index if multiple events happen simultaneously.
+        if merge_strategy=="first":
+            one_event_filtered = one_event.loc[one_event.groupby(pks)[event_ref].idxmin()]
```

Comment:
I did a some minimal tests and it looks like first/last can be negligibly to significantly faster than idxmin/idxmax.
```python
one_event_filterd =  one_event.groupby(pks).last().reset_index()
```
The reset_index forces the pks back to columns, so _can_ reorder columns.

And when trying to verify they got identical results, it turns out idxm**() can duplicate rows if the index is not unique().  Not sure if we should move the ask for a guardrail to an issue or just add a reset_index(drop=True) near the beginning of the merge code to prevent it; though the change to pd methods prevents this occurrence

Reply:
Switched the code to use first and last in f23dac4